### PR TITLE
fprotect: Fix bug when multiple boot stages are present (ACL only)

### PR DESCRIPTION
**Pull request by oyvindronningstad**
If a bootloader uses fprotect, the spent ACL region weren't accounted
for if e.g. the app tried to protect something.

Also add regression tests for this bug in the negative fprotect tests.
Refactor the test to test both the areas the immutable bootloader should
protect, and areas protected by the test itself.

Ref: NCSDK-10164

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

### CODE REVIEW

**Review by Vge0rge — APPROVED** [review]

File: tests/drivers/fprotect/negative/src/main.c [hunk 1]
```diff
@@ -39,67 +47,70 @@ void k_sys_fatal_error_handler(unsigned int reason, const z_arch_esf_t *pEsf)
...
+	nrfx_nvmc_bytes_write(addr, buf, sizeof(buf));
 	zassert_unreachable("Should have BUS FAULTed before coming here.");
 }
 
-
-static void test_flash_write_protected_fails(void)
+static void flash_write_protected_unmodified(uint32_t addr)
 {
-	flash_write_protected_fails(INVALID_WRITE_ADDR, true);
+	uint8_t buf[256];
```
Comment (Vge0rge) [inline tests/drivers/fprotect/negative/src/main.c:65]: I think that it's better to use BUF_SIZE here.

File: tests/drivers/fprotect/negative/src/main.c [hunk 2]
```diff
@@ -39,67 +47,70 @@ void k_sys_fatal_error_handler(unsigned int reason, const z_arch_esf_t *pEsf)
...
+#ifdef CONFIG_SECURE_BOOT
+	flash_write_protected_fails(TEST_FPROTECT_BOOTLOADER_PROTECTED, true);
+#endif
+}
 
 static void test_flash_read_protected_fails_r(void)
 {
-#ifdef PM_HW_UNIQUE_KEY_PARTITION_ADDRESS
+#ifdef CONFIG_HAS_HW_NRF_ACL
 	uint8_t buf[256];
```
Comment (Vge0rge) [inline tests/drivers/fprotect/negative/src/main.c:97]: And here, probably better to use BUF_SIZE.

Comment (Vge0rge) [inline tests/drivers/fprotect/negative/src/main.c:97]: And here, probably better to use BUF_SIZE.

**Review by Vge0rge — APPROVED** [review]

File: lib/fprotect/fprotect_acl.c [hunk 3]
```diff
@@ -8,17 +8,34 @@
 #include <hal/nrf_ficr.h>
 #include <errno.h>
 #include <sys/__assert.h>
+#include <kernel.h>
 
-static uint32_t region_idx;
+static int find_region_idx(uint32_t *region_idx)
```
Comment (sigvartmh) [inline lib/fprotect/fprotect_acl.c:13]: This function(correct me if I'm wrong) finds non locked regions right?
so maybe name it something with regards to that?

Comment (oyvindronningstad) [reply lib/fprotect/fprotect_acl.c]: I settled on find_free_region()

**Review by sigvartmh — APPROVED** [review]

**Review by frkv — APPROVED** [review]
LGTM!

File: lib/fprotect/fprotect_acl.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply lib/fprotect/fprotect_acl.c]: I think you were looking at an outdated diff?

**Review by sigvartmh — APPROVED** [review]
